File: src/Phan_05/polymorphism.cpp
#include <iostream>
using namespace std;

// Abstract Class
class Animal
{
public:
	// virtual functions
	virtual void show()
	{
		cout << "This is an animal\n";
	}
	// pure virtual functions
	virtual void speak() = 0;
	// pure destructor
	virtual ~Animal()
	{
		cout << "Animal destructor\n";
	}
};

class Dog : public Animal
{
public:
	void show()
	{
		cout << "This is a dog\n";
	}
	void speak()
	{
		cout << "Woof Woof\n";
	}
	~Dog()
	{
		cout << "Dog destructor\n";
	}
};

class Cat : public Animal
{
};

int main()
{
	// Animal animal;		// error: Abstract Class
	Dog dog; // success: Not an Abstract Class
	// Cat cat;				// error: Abstract Class
	Animal *animal1 = &dog; // success: use pointer class Animal

	animal1->show(); // This is a dog -> successfully

	// Test pure destructor
	Animal *animal2 = new Dog;
	delete animal2; // Dog destructor
					// Animal destructor

	return 0;
}

File: src/Phan_05/polymorphism.py
# 1. Inbuilt polymorphic functions
# len for string
print(len("Duy Onix"))  # 8

# len for tuple
print(len((1, 2, 3)))  # 3

# 2. User-defined polymorphic functions
def plus(a, b, c=0):  # c = 0: default value
    return a + b + c


print(plus(1, 2))  # 3
print(plus(1, 2, 3))  # 6

# 3. Polymorphism with class methods
class Dog:
    def speak(self):
        print("Woof Woof")


class Cat:
    def speak(self):
        print("Meow Meow")


dog = Dog()
cat = Cat()

# loop through each instance of Dog and Cat
for animal in (dog, cat):
    animal.speak()
    # Woof Woof
    # Meow Meow

# 4. Polymorphism with Inheritance
class Animal:
    def description(self):
        print("This is an animal")


class Dog_Inheritance(Animal):
    def description(self):
        print("This is a dog")


class Cat_Inheritance(Animal):
    def description(self):
        print("This is a cat")


animal = Animal()
dog = Dog_Inheritance()
cat = Cat_Inheritance()

animal.description()  # This is an animal
dog.description()  # This is a dog
cat.description()  # This is a cat

# 5. Polymorphism with a Function and objects

# function that can pass any type of params
def polymorphism(obj):
    obj.speak()


dog = Dog()
cat = Cat()

polymorphism(dog)  # Woof Woof
polymorphism(cat)  # Meow Meow
